rules/Transform/Rector/Class_/AddAllowDynamicPropertiesAttributeRector.php

```diff
@@ -0,0 +1,134 @@
+ */
+final class AddAllowDynamicPropertiesAttributeRector extends AbstractRector implements MinPhpVersionInterface
+{
+    /**
+     * @var string
+     */
+    private const ATTRIBUTE = 'AllowDynamicProperties';
+
+    public function __construct(
+        private FamilyRelationsAnalyzer $familyRelationsAnalyzer,
+        private PhpAttributeAnalyzer $phpAttributeAnalyzer,
+        private PhpAttributeGroupFactory $phpAttributeGroupFactory,
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Add the `AllowDynamicProperties` attribute to all classes', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+class SomeObject {
+    public string $someProperty = 'hello world';
+}
+CODE_SAMPLE
+
+                ,
+                <<<'CODE_SAMPLE'
+#[AllowDynamicProperties]
+class SomeObject {
+    public string $someProperty = 'hello world';
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+
+    public function getNodeTypes(): array
+    {
+        return [Class_::class];
+    }
+
+    /**
+     * @param Class_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (
+            $this->isDescendantOfStdclass($node) ||
+            $this->hasNeededAttributeAlready($node) ||
+            $this->hasMagicSetMethod($node)
+        ) {
+            return null;
+        }
+
+        return $this->addAllowDynamicPropertiesAttribute($node);
+    }
+
+    public function provideMinPhpVersion(): int
+    {
+        return PhpVersionFeature::DEPRECATE_DYNAMIC_PROPERTIES;
+    }
+
+    private function isDescendantOfStdclass(Class_ $node): bool
+    {
+        if (!$node->extends instanceof FullyQualified) {
+            return false;
+        }
+
+        $ancestorClassNames = $this->familyRelationsAnalyzer->getClassLikeAncestorNames($node);
+        return in_array('stdClass', $ancestorClassNames);
+    }
+
+    private function hasNeededAttributeAlready(Class_ $node): bool
+    {
+        $nodeHasAttribute = $this->phpAttributeAnalyzer->hasPhpAttribute($node, self::ATTRIBUTE);
+        if ($nodeHasAttribute === true) {
+            return true;
+        }
+
+        if (!$node->extends instanceof FullyQualified) {
+            return false;
+        }
+
+        // TODO: recursive check for if ancestors applied the attribute
+        $parentNode = $this->getParentClassNode($node);
+        return $this->hasNeededAttributeAlready($parentNode);
+    }
+
+    private function hasMagicSetMethod(Class_ $node): bool
+    {
+        $nodeSetMethod = $node->getMethod('__set');
+        if ($nodeSetMethod instanceof ClassMethod) {
+            return true;
+        }
+
+        if (!$node->extends instanceof FullyQualified) {
+            return false;
+        }
+
+        // TODO: recursive check for if ancestors defined __set
+        $parentNode = $this->getParentClassNode($node);
```

Comment:
Same note applies here - we'd need a valid way to get the node of the current node's parent class.
Once we can get that and pass it back recursively we can be sure it's checking all ancestors.

Reply:
You may verify it via `ClassReflection`, eg:

```php
use PHPStan\Reflection\ClassReflection;
use PHPStan\Reflection\ReflectionProvider;

    public function __construct(
        private ReflectionProvider $reflectionProvider
    ) {
    }


// ...
$classReflection = $this->reflectionProvider->getClass($className);
if ($classReflection->hasMethod('__set')) {

}
```